### ISSUE TITLE
Prisma hangs when javascript is minified

### ISSUE THREAD

**Issue by mdesousa** [prisma/prisma#12355]
### Bug description

We noticed that if code is bundled and prisma throws a constraint error (e.g. duplicate key on insert, or invalid foreign key) the execution hangs.

### How to reproduce

Implement code that would cause prisma to throw a constraint error. If executed, the exception is properly raise.
Now bundle the code with a bundler (e.g. esbuild). Execute again and notice that the application hangs.

### Expected behavior

The bundled code should throw the expected exception

### Prisma information

The issue seems to be unrelated to the schema. Just need a couple of tables with a FK constraint.

### Environment & setup

- OS: AWS Lambda (Amazon Linux 2)
- Database: PostgreSQL
- Node.js version: 14

### Prisma Version

```
prisma                  : 3.10.0
@prisma/client          : 3.10.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 73e60b76d394f8d37d8ebd1f8918c79029f0db86 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : 73e60b76d394f8d37d8ebd1f8918c79029f0db86
Studio                  : 0.458.0
```

**Comment by janpio** [comment]
Can you maybe provide a reproduction repository with an example for this? Would be very helpful to speed up our reproduction. Thanks.

**Comment by mdesousa** [comment]
hi @janpio , please take a look at this repository: https://github.com/holasivo/prisma-12355
in testing with this simple repo, the code actually works... but it is very slow when minified
i'm wondering now if the issue is related to the exception including the source code. trying to print that may be a lot slower for minified code (there are no line breaks, etc.). our application has a lot more code getting minified and it times out after 30 seconds.
i'd also recommend to avoid printing source code in the exception message for security reasons... the stack trace already provides the code context.
thanks!

**Comment by janpio** [comment]
I can confirm the minified version is significantly slower:

```
PS C:\Users\Jan\Documents\throwaway\12355> Measure-Command { npm start }
PrismaClientKnownRequestError:
Invalid `prisma.employee.create()` invocation in
C:\Users\Jan\Documents\throwaway\12355\src\index.ts:5:25

  2
  3 const main = async () => {
  4   const prisma = new PrismaClient();
→ 5   await prisma.employee.create(
  Foreign key constraint failed on the field: `Employee_department_id_fkey (index)`
    at Object.request (C:\Users\Jan\Documents\throwaway\12355\node_modules\@prisma\client\runtime\index.js:39818:15)
    at async PrismaClient._request (C:\Users\Jan\Documents\throwaway\12355\node_modules\@prisma\client\runtime\index.js:40646:18)
    at async main (C:\Users\Jan\Documents\throwaway\12355\src\index.ts:5:3) {
  code: 'P2003',
  clientVersion: '3.11.0',
  meta: { field_name: 'Employee_department_id_fkey (index)' }
}


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 3
Milliseconds      : 888
Ticks             : 38882102
TotalDays         : 4.50024328703704E-05
TotalHours        : 0.00108005838888889
TotalMinutes      : 0.0648035033333333
TotalSeconds      : 3.8882102
TotalMilliseconds : 3888.2102
```


vs.

```
PS C:\Users\Jan\Documents\throwaway\12355> Measure-Command { npm run start:esb }
PrismaClientKnownRequestError:
Invalid `ft.PrismaClient().employee.create()` invocation in
C:\Users\Jan\Documents\throwaway\12355\dist\index-esb.js:314:93036

  311     }
  312   }
  313 })`);let s=a?{"X-DEBUG-FATAL":"1"}:{"X-DEBUG-NON-FATAL":"1"};return this._request({action:"queryRaw",args:{query:"SELECT 1",parameters:void 0},clientMethod:"queryRaw",dataPath:[],runInTransaction:!1,headers:s,callsite:getCallSite(this._errorFormat)})}_transactionWithArray(a){let s=this._transactionId++,i=getLockCountPromise(a.length),o=a.map(p=>{var n;if(p?.[Symbol.toStringTag]!=="PrismaPromise")throw new Error("All elements of the array need to be Prisma Client promises. Hint: Please make sure you are not awaiting the Prisma client calls you intended to pass in the $transaction function.");return(n=p.requestTransaction)==null?void 0:n.call(p,s,i)});return Promise.all(o)}async _transactionWithCallback(a,s){let i=await this._engine.transaction("start",s),o;try{o=await a(transactionProxy(this,i.id)),await this._engine.transaction("commit",i)}catch(p){throw await this._engine.transaction("rollback",i).catch(()=>{}),p.clientVersion=this._clientVersion,p}return o}async $transaction(a,s){return this._hasPreviewFlag("interactiveTransactions")?typeof a=="function"?this._transactionWithCallback(a,s):this._transactionWithArray(a):this._transactionWithArray(a)}async _request(a){this._hasPreviewFlag("tracing")||delete a.otelCtx;try{let s={args:a.args,dataPath:a.dataPath,runInTransaction:a.runInTransaction,action:a.action,model:a.model},i=-1,o=__name2(p=>{let n=this._middlewares.query.get(++i);if(n)return n(p,o);let l={...a,...p};return this._executeRequest(l)},"consumer");return await new import_async_hooks.AsyncResource("prisma-client-request").runInAsyncScope(()=>runInChildSpan("request",a.otelCtx,()=>o(s)))}catch(s){throw s.clientVersion=this._clientVersion,s}}async _executeRequest({args:a,clientMethod:s,dataPath:i,callsite:o,runInTransaction:p,action:n,model:l,headers:d,transactionId:u,otelCtx:m,lock:c,unpacker:f}){let _,h=actionOperationMap[n];(n==="executeRaw"||n==="queryRaw"||n==="runCommandRaw")&&(_=n);let b;if(l){if(b=this._dmmf.mappingsMap[l],!b)throw new Error(`Could not find mapping for model ${l}`);_=b[n]}if(h!=="query"&&h!=="mutation")throw new Error(`Invalid operation ${h} for action ${n}`);let g=this._dmmf.rootFieldMap[_];if(!g)throw new Error(`Could not find rootField ${_} for action ${n} for model ${l} on rootType ${h}`);let{isList:y}=g.outputType,w=getOutputTypeName(g.outputType.type),v=getRejectOnNotFound(n,w,a,this._rejectOnNotFound),S=makeDocument({dmmf:this._dmmf,rootField:_,rootTypeName:h,select:a});if(S.validate(a,!1,s,this._errorFormat,o),S=S,import_debug8.default.enabled("prisma:client")){let E=String(S);debug8("Prisma Client call:"),debug8(`prisma.${s}(${printJsonWithErrors({ast:a,keyPaths:[],valuePaths:[],missingItems:[]})})`),debug8("Generated request:"),debug8(E+`
→ 314 `)}return d=applyTracingHeaders(d,m),await c,this._fetcher.request(
  Foreign key constraint failed on the field: `Employee_department_id_fkey (index)`
    at Object.request (C:\Users\Jan\Documents\throwaway\12355\dist\index-esb.js:296:56)
    at async r._request (C:\Users\Jan\Documents\throwaway\12355\dist\index-esb.js:313:1493)
    at async _t (C:\Users\Jan\Documents\throwaway\12355\dist\index-esb.js:314:92999) {
  code: 'P2003',
  clientVersion: '3.11.0',
  meta: { field_name: 'Employee_department_id_fkey (index)' }
}


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 8
Milliseconds      : 220
Ticks             : 82207279
TotalDays         : 9.51473136574074E-05
TotalHours        : 0.00228353552777778
TotalMinutes      : 0.137012131666667
TotalSeconds      : 8.2207279
TotalMilliseconds : 8220.7279
```

**Comment by janpio** [comment]
No information about that, sorry. But as it has been confirmed already, we should hopefully be able to tackle it sooner than all the bugs that are not confirmed.

**Comment by mdesousa** [comment]
hi @janpio @dpetrick , just noticed that the priority was set to low :(
this is an important performance issue for running prisma on lambda. is there at least a way to opt-out of the verbose error messages generated by prisma? that would allow us to work around the problem. thanks.

**Comment by millsp** [comment]
Because we need to find the schema and the engine in order to run the queries, we have logic to find that but it is suboptimal.

**Comment by buffolander** [comment]
Try removing this line https://github.com/holasivo/prisma-12355/blob/bcdd7317a0ee4ba4c44b4ee14cedd2b55d14587b/tsconfig.json#L16

**Comment by SevInf** [comment]
Here is my investigation:

1. The issue does not reproduce for me anymore on 4.0.0+
2. Culprit was a regexp of accidental exponential complexity we use within our error rendering code. In this case, stars aligned just enough so that bundled code would actually trigger exponential backtracking. 
3. Recommendation to anyone experiencing the issue: update to 4.x. Code, that triggers backtracking is not there anymore. If you can't - set `NODE_ENV` environment variable to `production`, either when you run the app or during the build. For esbuild, used in this reproduction it can be done adding `--define:process.env.NODE_ENV='production'` switch.
4. Fix for regexp itself is in #17907

**Comment by mdesousa** [comment]
thanks @SevInf ... just upgrade the repo to the latest packages and now it fails instantly!
resolving...
